service/history/task/cross_cluster_task.go

```diff
@@ -37,136 +40,146 @@ import (
 		processingState processingState
 		priority        int
 		attempt         int
-		timeSource      clock.TimeSource
-		submitTime      time.Time
-		logger          log.Logger
-		eventLogger     eventLogger
-		scopeIdx        int
-		scope           metrics.Scope
-		taskExecutor    Executor        //TODO: wire up the dependency
-		taskProcessor   Processor       //TODO: wire up the dependency
-		redispatchFn    func(task Task) //TODO: wire up the dependency
-		maxRetryCount   dynamicconfig.IntPropertyFn
-		settable        future.Settable
+
+		shard          shard.Context
+		timeSource     clock.TimeSource
+		submitTime     time.Time
+		logger         log.Logger
+		eventLogger    eventLogger
+		scopeIdx       int
+		scope          metrics.Scope
+		taskExecutor   Executor
+		resubmitTaskFn func(task Task)
+		maxRetryCount  int
 	}
 )
 
-// NewCrossClusterTaskForTargetCluster creates a CrossClusterTask
-// for the processing logic at target cluster
-// the returned the Future will be unblocked when after the task
-// is processed. The future value has type types.CrossClusterTaskResponse
-// and there will be not error returned for this future. All errors will
-// be recorded by the FailedCause field in the response.
-func NewCrossClusterTaskForTargetCluster(
-	shard shard.Context,
-	taskRequest *types.CrossClusterTaskRequest,
-	logger log.Logger,
-	maxRetryCount dynamicconfig.IntPropertyFn,
-) (CrossClusterTask, future.Future) {
-	// TODO: create CrossClusterTasks based on request
-	// for now create a dummy CrossClusterTask for testing
-	future, settable := future.NewFuture()
-	return &crossClusterSignalWorkflowTask{
-		crossClusterTaskBase: &crossClusterTaskBase{
-			Info: &persistence.CrossClusterTaskInfo{
-				TaskID: taskRequest.TaskInfo.TaskID,
-			},
-			settable: settable,
-		},
-	}, future
-}
-
-// NewCrossClusterSignalWorkflowTask initialize cross cluster signal workflow task and task future
-func NewCrossClusterSignalWorkflowTask(
+// NewCrossClusterSourceTask creates a cross cluster task
+// for the processing it at the source cluster
+func NewCrossClusterSourceTask(
 	shard shard.Context,
 	taskInfo Info,
+	taskExecutor Executor,
 	logger log.Logger,
-	timeSource clock.TimeSource,
+	resubmitTaskFn func(task Task),
 	maxRetryCount dynamicconfig.IntPropertyFn,
-) CrossClusterTask {
-	crossClusterTask := newCrossClusterTask(
-		shard,
-		taskInfo,
-		logger,
-		timeSource,
-		maxRetryCount)
-	return &crossClusterSignalWorkflowTask{
-		crossClusterTask,
+) Task {
+	return &crossClusterSourceTask{
+		crossClusterTaskBase: newCrossClusterTaskBase(
+			shard,
+			taskInfo,
+			taskExecutor,
+			logger,
+			shard.GetTimeSource(),
+			resubmitTaskFn,
+			maxRetryCount,
+		),
 	}
 }
 
-// NewCrossClusterCancelWorkflowTask initialize cross cluster cancel workflow task and task future
-func NewCrossClusterCancelWorkflowTask(
+// NewCrossClusterTaskForTargetCluster creates a cross cluster task
+// for the processing it at the target cluster
```

Comment:
This function is called at the target all the time right? Can you change the comment like 

```
// NewCrossClusterTaskForTargetCluster is called at the target cluster to 
// process the cross cluster task
```

Confusion in the original comment is that we don't know if we are going to send it from the source or if we are already at the target cluster.

Reply:
nit: Similarly for the name, you might want to change it to `NewCrossClusterTargetTask`

Reply:
Done.